sssd_test_framework/utils/dbus_types.py

```diff
@@ -0,0 +1,675 @@
+    "DBUSTypeDict",
+    "DBUSTypeVariant",
+    "DBUSSignatureReader",
+]
+
+
+# A NOTE ABOUT DBUS-SEND
+# **********************
+# It is important to know that dbus-send does not have a symmetric
+# behavior with relation to how the types are described in text
+# form when used as parameters or results.
+
+
+# TYPES PENDING IMPLEMENTATION
+# ****************************
+# The following D-Bus types are not implemented because they are not
+# currently recognized by dbus-send:
+# - Signature
+# - UNIX FD
+# - Struct
+
+
+class MutableString:
+    """A helper class implementing an extremely simple mutable string"""
+
+    def __init__(self, text: str):
+        self.text = text
+
+    def __str__(self):
+        return self.text
+
+
+class DBUSResult(MutableString):
+    """
+    The result from the D-Bus operation executed by dbus-send.
+    It is used to simplify parsing of the result as the string
+    must be modified and have no carriage return.
+    """
+
+    def __init__(self, text: str):
+        self.text = text.replace("\n", "")
+
+
+class DBUSType(ABC):
+    """
+    Abstract class to create classes that handle type conversion
+    between D-Bus and Python.
+    """
+
+    # It would be good to define here the following abstract class members:
+    #    _type_prefix
+    #
+    # But there is not such thing as _abstract class members_ in Python and
+    # linters will complain although it will work because the contrete
+    # subclasses will declare the concrete member. That's why we are ignoring
+    # those errors.
+
+    @property
+    @abstractmethod
+    def value(self):
+        """The Python value"""
+        pass
+
+    @value.setter
+    @abstractmethod
+    def value(self, val):
+        pass
+
+    def __str__(self) -> str:
+        """
+        Produce a string representation for the instantiated type's Python value.
+        """
+        return str(self._value)  # type: ignore[attr-defined]
+
+    def clone(self):
+        """Clone this object. Structure is deepcopied, but values are not."""
+        return self.__class__()
+
+    def param(self) -> str:
+        """
+        Produce a string representation for the instantiated type suitable for
+        using as a parameter for ``dbus-send``.
+        """
+        return f"{self._type_prefix}:{self}"  # type: ignore[attr-defined]
+
+    @abstractmethod
+    def parse(self, result: DBUSResult) -> None:
+        """
+        Parse a resulting string from ``dbus-send`` and instantiate
+        the corresponding ``DBUSType``.
+
+        It consumes the beginning of the provided result.
+        """
+        raise NotImplementedError("Abstract method invoked")
+
+    @classmethod
+    def _guess_next_type(cls, result: DBUSResult) -> DBUSType:
+        """
+        Try to guess which is the next type.
+
+        This is mostly needed for variants whose signature doesn't include the subtype.
+        """
+        for subcls in cls.__subclasses__():
+            try:
```

Comment:
The call to `subcls._guess_next_type(result)` can also throw a `TypeError` exception that I want to skip.

I find cleaner the code this way, in particular because the other invocations (like `len()`, `__subclass__()` and `subcls()`), I believe, will not throw this exception. With these changes the code would become:

```
        for subcls in cls.__subclasses__():
            if len(subcls.__subclasses__()) == 0:
                obj = subcls()
                try:
                    obj.parse(DBUSResult(result.text))
                except TypeError:
                    continue
            else:
                try:
                    obj = subcls._guess_next_type(result)
                except TypeError:
                    continue
            return obj

        raise TypeError("Couldn't guess the type")
```

Do you still want me to change it?

Reply:
Ok, keep it as is.